### ISSUE TITLE
Add a Golang core scanner to ClairCore

### ISSUE THREAD

**Issue by jasinner** [quay/claircore#127]
Description Feature Request (originally raise [in clair](https://github.com/quay/clair/issues/920))

Add a 'Golang core' scanner to Clair. Because Golang binaries are statically compiled it can be hard to track and fix security vulnerabilities which are present in them. This feature aims to improve the situation by identifying Golang binaries in a layer, by using Golang compiler version the binary was compiled with, and matching that to NIST NVD data using the `cpe:2.3:a:golang:go:*:*:*:*:*:*:*:*` cpe.

A distribution scanner should be added which inspects every file in a docker layer using the linux tools, 'file', and 'gdb' to determine the Golang compiler version used to compile the binary.

`file <target-file> | grep ELK`
`gdb -q -ex "p 'runtime.buildVersion'" -ex quit <target-file> 2>/dev/null`

A new vulnsrc should be added which uses NIST NVD 1.1 Json data, filtered by Golang CPE. In that data, it's possible to identify which version addresses the vulnerability. For example CVE-2015-5741 affected golang up to, but excluding 1.4.3, from the 2015 data.

```
{
    "cve" : {
      "data_type" : "CVE",
      "data_format" : "MITRE",
      "data_version" : "4.0",
      "CVE_data_meta" : {
        "ID" : "CVE-2015-5741",
        "ASSIGNER" : "cve@mitre.org"
      },
      "problemtype" : {
        "problemtype_data" : [ {
          "description" : [ {
            "lang" : "en",
            "value" : "CWE-444"
          } ]
        } ]
      },
... 
   "configurations" : {
      "CVE_data_version" : "4.0",
      "nodes" : [ {
        "operator" : "OR",
        "cpe_match" : [ {
          "vulnerable" : true,
          "cpe23Uri" : "cpe:2.3:a:golang:go:*:*:*:*:*:*:*:*",
          "versionEndExcluding" : "1.4.3"
        } ]
      }
```

**Comment by ldelossa** [comment]
@jasinner 

I like this a lot. And I'm definitely into supporting this. 

My one concern is starting a gdp process for every file in a tar ball. This is expensive. Do you think we can come up with a way to do the 'runtime.Buildversion' symbol check in memory? Maybe a library exists that can parse elf format in memory?

**Comment by jasinner** [comment]
@ldelossa Ok, what if we used https://golang.org/pkg/debug/elf/#File.Symbols to read 'buildVersion' symbol instead of gdb?

**Comment by ldelossa** [comment]
@jasinner looks promising. Let me test out this library

**Comment by jjnoakes** [comment]
Just wanted to chime in with a +1 and offer some help. I have some 'FROM scratch' images that contain only a single go binary and Clair reports no detected features which appears indistinguishable from an unknown OS, but the two have very different security profiles and surface area, and I think anything which allows Clair to distinguish the two and report them differently would be a big win. If there's work begin done on this front and I can provide any help, let me know. (Is there a branch or fork anywhere?)

**Comment by jasinner** [comment]
Started something here:
https://github.com/jasinner/claircore/tree/master/golang

**Comment by ldelossa** [comment]
@jasinner @hdonnay has began working on a golang scanner here: 
https://github.com/quay/claircore/pull/363

We are currently waiting on Go's official security database to come online. Are you aware of this announcement?

**Comment by jasinner** [comment]
I see the Go community have started publishing entries in the official security database here:

https://storage.googleapis.com/go-vulndb/

**Comment by bored-engineer** [comment]
The vuln DB is live: https://go.dev/security/vuln/database including an official Go client: https://pkg.go.dev/golang.org/x/vuln/client

Additionally, Golang has added the [debug/buildinfo.Read](https://pkg.go.dev/debug/buildinfo#Read) function which can extract all modules and versions from a Golang binary, which should make the implementation much easier.

**Comment by bored-engineer** [comment]
I started writing a `Scanner` implementation using `buildinfo.Read` based on how I see the other scanners are implemented: https://github.com/quay/claircore/compare/v1.4.11...bored-engineer:claircore:golang?expand=1 

I did not include an `Updater` implementation as I think it can just be inherited that from the OSV implementation @hdonnay started (but is not enabled yet) as long as the packages produced by the scanner are correct.

Unfortunately, I ran out of time to actually validate it works and write tests and likely won't be able to re-visit it. Feel free to pickup any of the code from my fork if it looks like a useful starting point.
